---
title: "class15"
author: "Pratik Varade"
date: "11/14/2019"
output: github_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```


Data from : 
Trapnell C, Hendrickson DG, Sauvageau M, Goff L et al. "Differential analysis of gene regulation at transcript resolution with RNA-seq". Nat Biotechnol 2013 Jan;31(1):46-53. PMID: 23222703



Section 1:
```{r}
library(DESeq2)
```



Load up our files
```{r}
metaFile <- "GSE37704_metadata.csv"
countFile <- "GSE37704_featurecounts.csv"

#Import meta data and take a peak
colData = read.csv(metaFile, row.names=1)
head(colData)
```

```{r}
# Import countdata
countData = read.csv(countFile, row.names=1)
head(countData)
```

We need to remove the odd first $length col. This is because that row is going to cause problems with the required matching to the metadata file.
```{r}
# Note we need to remove the odd first $length col
countData <- as.matrix(countData[,-1])
head(countData)
```
 
 Double check that the colnames in the countData match the id values in the Col Data metadata file.
 
```{r}
colnames(countData)

```

```{r}
rownames(colData)
```
You could also use this. All makes sure that ALL of the outputs are the same.
```{r}
all(colnames(countData) == rownames(colData))
```

##Remove zero count genes 

We want to remove genes that have 0 count values in all experiments ( rows that have 0 across all cols)
```{r}
countData =countData[rowSums(countData)!=0, ]
head(countData)
```

#DESeq Analysis (already ran library(DESeq above)
```{r}
#Setup the object with our data in the way DESeq wants it
dds = DESeqDataSetFromMatrix(countData=countData,
                             colData=colData,
                             design=~condition)

#Run the analysis
dds = DESeq(dds)
```

Get our results
```{r}
res = results(dds)
res
```

```{r}
plot( res$log2FoldChange, -log(res$padj) )

```

```{r}
mycols <- rep("gray", nrow(res) )

# Color red the genes with absolute fold change above 2
mycols[ abs(res$log2FoldChange) > 2 ] <- "red"

# Color blue those with adjusted p-value less than 0.01
#  and absolute fold change more than 2
inds <- (res$padj < 0.01) & (abs(res$log2FoldChange) > 2 )
mycols[ inds ] <- "blue"

plot( res$log2FoldChange, -log(res$padj), col=mycols, xlab="Log2(FoldChange)", ylab="-Log(P-value)" )
```
#Do in Console

```{r, }
BiocManager::install("AnnotationDbi")
BiocManager::install("org.Hs.eg.db")
```
Load the packages now
```{r}
library("AnnotationDbi")
library("org.Hs.eg.db")
```



```{r}
columns(org.Hs.eg.db)
```

```{r}
res$symbol = mapIds(org.Hs.eg.db,
                    keys=row.names(countData), #where are my IDS
                    keytype="ENSEMBL",         #what format are my IDs
                    column="SYMBOL",          #the new format I want
                    multiVals="first")
res

```

```{r}
res$entrez = mapIds(org.Hs.eg.db,
                    keys=row.names(countData), 
                    keytype="ENSEMBL",         
                    column="ENTREZID",         
                    multiVals="first")
```


##Section 2: Pathway Analysis

Here we are going to use the gage package for pathway analysis. Once we have a list of enriched pathways, we're going to use the pathview package to draw pathway diagrams, shading the molecules in the pathway by their degree of up/down-regulation.

```{r}
library(pathview)
library(gage)
library(gageData)
```

```{r}
data(kegg.sets.hs)
data(sigmet.idx.hs)

# Focus on signaling and metabolic pathways only
kegg.sets.hs = kegg.sets.hs[sigmet.idx.hs]

# Examine the first 3 pathways
head(kegg.sets.hs, 3)
```




```{r}
foldchanges = res$log2FoldChange
names(foldchanges) = res$entrez
head(foldchanges)

```

```{r}
keggres = gage(foldchanges, gsets=kegg.sets.hs)
#now lets look at the object returned from gage()
attributes(keggres)
```





# Look at the first few down (less) pathways
```{r}
head(keggres$less)
```


This makes a pathway plot with our RNA-Seq expression.

```{r}
pathview(gene.data=foldchanges, pathway.id="hsa04110")

```

 A different PDF based output of the same data

```{r}
pathview(gene.data=foldchanges, pathway.id="hsa04110", kegg.native=FALSE)
```

Section 3 : Gene Ontology (GO)
```{r}
data(go.sets.hs)
data(go.subs.hs)

# Focus on Biological Process subset of GO
gobpsets = go.sets.hs[go.subs.hs$BP]

gobpres = gage(foldchanges, gsets=gobpsets, same.dir=TRUE)

lapply(gobpres, head)
```

























